### DISCUSSION TITLE
Typesafe Custom Error Handling

### DISCUSSION THREAD

**Discussion by jonathangpk** [trpc/trpc#3300] [Q&A] [answered]
I am pretty new to trpc and was wondering if there is a suggested way to explicitly define Errors for each response so that the client knows all errors for a router and can handle them exhaustively. I didn't find what I was looking for in the docs.

For example, let's say we have authentication endpoints. 

```ts
const authRouter = router({
  login: ...,
  register: ...,
})
```
We also have some Errors defined for these routes:
```ts
class AuthError extends Error {}

// Register
class EmailAlreadInUse extends AuthError {}
class InvalidEmail extends AuthError {}
class PasswordNotStrongEnough extends AuthError {}

// Login
class InvalidPasswordOrEmail extends AuthError {}
class TooManyLoginAttempts extends AuthError {}
```

Is there a suggested way of configuring them to let the client know what errors a route can throw? And make sure the client can check it?
To my knowledge, typescript cannot infer yet what kind of Errors a function throws. So using `throw new Error()` might not work for this. 
My current approach would be with returning objects representing errors instead of throwing them. Is that something that is already supported in trpc? Or is there a suggested pattern?

My approach with returning Error objects looks like this:
```ts
interface CustomError {
  code: string
  message?: string
}
interface RegistrationError extends CustomError {
  code: 'EmailAlreadInUse' | 'InvalidEmail' | 'PasswordNotStrongEnough';
}
interface LoginError extends CustomError {
  code: 'InvalidPasswordOrEmail' | 'TooManyLoginAttempts';
}
```
Then a API router would return a result or a error 
```ts
type Response<Data, Error extends CustomError> = {
  type: 'error';
  error: Error;
} | {
  type: 'result';
  result: Data;
};
// Error could also be an array here, but simpler for this example

// Symbolic for the response of authRouter.login
function login(input: LoginInput): Response<User, LoginError> {}
```
On the client side we can now handle all the errors in a way that looks something like this.
```ts
const response = trpc.authRouter.login(loginInput)
if (response.errors) {
  // Handle Unexpected Errors or TRPCError
}
if (response.data) {
    if (response.data.type === 'result') {
      // got results
    } else { // response.data.type === 'error'
      // Handle Exepted Errors
      switch(response.data.error.code) {
        case: 'InvalidPasswordOrEmail': 
        // handle this
        case: 'TooManyLoginAttempts':
        // handle this
    }
    }    
}
```
When using this in components now, we have this duplicate of `response.data.results` and we have two potential error objects:
- `response.error` for unexpected errors
- `response.data.error` for expected errors

Is there a way to combine them that is suggested by trpc?

Also interested to hear how others are dealing with this.

**Comment by KATT** [comment, marked as answer]
> Is there a suggested way of configuring them to let the client know what errors a route can throw? And make sure the client can check it?

No. We did play with it for a while when doing the design for v10 but we abandoned it because errors are notoriously hard to predict.

However, you can do this on a router level. See here for an example of how we can infer a zod errors: https://trpc.io/docs/error-formatting

I would advise against trying to send an instance of a class over the wire (although it is possible [with superjson](https://github.com/blitz-js/superjson#recipes)), do instead some sort of discriminated union.

```ts

const myKnownErrors = [AuthError, EmailAlreadyInUse];


type MyError = myKnownErrors[number];

function isMyError(val: unknown): val is MyError {
  return myKnownErrors.some(ErrorClass => val instanceof ErrorClass)
}

function formatError(err: MyError) {
  switch (err) {
   case AuthError:
      return {
        type: "AuthError" as const,
      } 
      // [...]
  }
}


export const t = initTRPC.context<Context>().create({
  errorFormatter({ shape, error }) {
    return {
      ...shape,
      data: {
        ...shape.data,
        ...isMyError(error.cause) ? formatError(err) : {},
      },
    };
  }
})

```

Another way is to make sure that all your classes `implement`s a specific interface that make them serializable, that would cut it down a bit:





```ts

const myKnownErrors = [AuthError, EmailAlreadyInUse];


type MyError = myKnownErrors[number];

function isMyError(val: unknown): val is MyError {
  return myKnownErrors.some(ErrorClass => val instanceof ErrorClass)
}

function formatError(err: MyError) {
  switch (err) {
   case AuthError:
      return {
        type: "AuthError" as const,
      } 
      // [...]
  }
}


export const t = initTRPC.context<Context>().create({
  errorFormatter({ shape, error }) {
    return {
      ...shape,
      data: {
        ...shape.data,
        ...isMyError(error.cause) ? error.cause.serialize() : {},
      },
    };
  }
})

```


:warning: I wrote all of this code in markdown without testing it, it's just to convey ideas.

**Comment by jonathangpk** [comment]
Thank you for the fast answer, I saw the global handling, but I was missing something on a route level.

> errors are notoriously hard to predict

agreed, that's why I was thinking of splitting it into expected / unexpected errors, types for expected, default types for unexpected types.

Thanks for the input, I'll probably stick with the `Response<Data, Error>` method :)

**Comment by CanRau** [comment]
Just stumbled upon this discussion here and [this issue](https://github.com/trpc/trpc/issues/3438) as I'm also confused and a little struggling as to how to handle this properly, so far I like your manual `: Response` casting, another way could be defining the procedure function beforehand, infer the response and then pass those to `publicProcedure.input(schema).mutation(myProcedure)`, haven't tried it yet though
